Add OpenCensus instrumentation for tracing

ocervell opened the pull request:
This PR brings support for OpenCensus tracing for all gRPC Forseti services:

* Add OpenCensus interceptors to Forseti's `gRPC` server and client
* Set up an exporter of traces to Stackdriver Trace (plan on supporting more via cfg changes)
* Add appropriate IAM role `cloudtrace.agent` to allow client/server VMs to write traces to Stackdriver (during deployment)
* Add 'opencensus' and 'google-cloud-trace' dependencies

ocervell commented:
Seems like the library fails to contact Stackdriver API on the Travis CI VM, do we have gcloud credentials set up on that VM ? 

Maybe we can just disable the Stackdriver exporter in testing mode and have the FileExporter that works locally instead.

blueandgold commented:
Relates to #1845, #1438

ocervell commented:
I signed it !

ahoying commented:
Please make this change aware of if the opencensus package is installed, if it is not installed it should be a no-op.

For changes like this, where a new third party library is introduced for an optional feature, or a feature that can be disabled without impacting the rest of forseti, the code needs to fail back cleanly when that library is not installed.

This happens to be the first new feature to be implemented where this would apply.

ocervell commented:
We're planning on using OpenCensus for our monitoring as well. 

While tracing could be made optional and configurable, monitoring (exposing OpenCensus metrics to monitoring systems like Stackdriver) seems required for Forseti, so the dependency would be as well.

Is there currently any way to add an optional dependency in Forseti deployments ?

mirons-google commented:
Since this is a first step in working to solve #1845 #1844 and #1438 I think it makes sense to consider the package as required.

The preferred way to do that is in setup.py currently.

ahoying commented:
Forseti is both a deployment but also a set of libraries that are used elsewhere. Introducing additional third party dependencies makes the use of forseti libraries more difficult when a full deployment is not desired.

I recommend wrapping the opencensus imports in the tracing module in a try catch, and setting a global to True or False depending on if they are available or not.

Then the code can handle the cases where opencensus is installed or not installed by checking the value of that global, and you can fall back gracefully.

ocervell commented:
@ahoying made some good points there. Users should be able to install the `forseti-security` library without extra dependencies. 

My last commit brings support for:

- More interceptors to be added (e.g a MetricsInterceptor)
- A server CLI flag `--enable-tracing` that's disabled by default.
- Added `--enable-tracing` to our `systemctl` startup script.
- Tracing libraries removed as required dependencies and added as `OPTIONAL_DEPENDENCIES` to `setup.py`. They can be installed using `pip install .[tracing]`, statement I've added to our Forseti deployments that will enable tracing by default.

Please review and let me know what you think.

blueandgold commented on google/cloud/forseti/services/utils.py:
```diff
@@ -258,3 +259,20 @@ def get_resource_id_from_type_name(type_name):
+        imp.find_module('opencensus')
+        return True
+    except ImportError:
+        logging.warning('Cannot enable tracing because the `opencensus` library\
```
Backslash should not be used for line cotinuations.
```
logging.warning(
    'Cannot enable tracing because the `opencensus` library'
    'was not found. Run `pip install .[tracing]` to install tracing libraries.')

blueandgold commented on deployment-templates/compute-engine/client/forseti-instance-client.py:
```diff
@@ -133,6 +133,9 @@ def GenerateConfig(context):
 pip install -q --upgrade -r requirements.txt
 
+# Install instrumentation libs
+pip install .[tracing]
```
This is new for me.  But is this relative import the preferred way here?  Would it be better to do specify it as:
```
pip install <module>[tracing]
```
https://stackoverflow.com/a/24640526/2830207

blueandgold commented on google/cloud/forseti/services/utils.py:
```diff
@@ -258,3 +259,20 @@ def get_resource_id_from_type_name(type_name):
     return type_name
+
+
+def opencensus_enabled():
```
This method's naming would be more accurate as `is_opencensus_installed()`, because that's what it's checking.

blueandgold commented on google/cloud/forseti/services/client.py:
```diff
@@ -797,3 +799,17 @@ def delete_model(self, model_name):
+            tuple: A tuple of interceptors.
+        """
+        interceptors = []
+        if opencensus_enabled():
```
What happens if `enable_tracing` is disabled on the server?  Is there any impact here in the client?

blueandgold replied:
Can we just leave a comment here?
```
# It's okay for this to be enabled on the client, even if the tracing flag is disabled on the server.
```

blueandgold commented on google/cloud/forseti/services/client.py:
```diff
@@ -797,3 +799,17 @@ def delete_model(self, model_name):
         return self.model.delete_model(model_name)
+
+    @property
+    def interceptors(self):
```
Why is this a property method here in the `services/client.py`, but `setup_interceptors()` in `services/server.py`?

blueandgold replied:
Got it, thanks.

blueandgold commented on google/cloud/forseti/services/server.py:
```diff
@@ -170,6 +196,10 @@ def main():
         help='Print log to console.')
+    parser.add_argument(
+        '--enable-tracing',
+        action='store_true',
```
Is this, and should be, enabled?  Per your PR description, this should be disabled by default?

ocervell replied:
Yes, `store_true` actually stores False by default in Python (cf: https://docs.python.org/2/library/argparse.html#action)

blueandgold replied:
TIL  :)

blueandgold commented on google/cloud/forseti/services/tracing.py:
```diff
@@ -0,0 +1,84 @@
+        host_port=endpoint)
+
+
+def trace_server_interceptor():
```
Method naming should generally start with verbs.  So, `create_trace_server_interceptor()`.

ocervell replied:
Ok. See previous comment to drop the `_trace`

blueandgold commented on google/cloud/forseti/services/tracing.py:
```diff
@@ -0,0 +1,84 @@
+LOGGER = logger.get_logger(__name__)
+
+
+def trace_client_interceptor(endpoint):
```
Method naming should generally start with verbs.  So, `create_trace_client_interceptor()`.

ocervell replied:
I'd simplify it to `create_client_interceptor` and `create_server_interceptor` then.

I think it looks good that way:
```
from google.cloud.forseti.services import tracing
tracing.create_client_interceptor(endpoint='localhost:5001')
tracing.create_server_interceptor()
```

When we do the monitoring piece, we'll have similarly:
```
from google.cloud.foresti.services import monitoring
monitoring.create_client_interceptor(endpoint='localhost:5001')
monitoring.create_server_interceptor()
```

Side question: should we move `tracing.py` (and future `monitoring.py`) into `google/cloud/forseti/common/` or does it make sense to keep it next to the `client.py` and `server.py` ?

blueandgold replied:
Your naming proposal sounds great!

I am neutral on moving this to the common package, as that to me is for sharing between different packages.  Do you think that will be the case in the future?

ocervell replied:
Yes, I think one of the goals after this PR is to implement some kind of decorator to enable more tracing (sub-spans) on certain functions / classes. In that case those would be shared by other services.

ocervell commented on google/cloud/forseti/services/tracing.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def trace_client_interceptor(endpoint):
+    """Intercept gRPC calls on client-side and add tracing information
```
No problem.

ocervell commented on google/cloud/forseti/services/tracing.py:
```diff
@@ -0,0 +1,84 @@
+        LOGGER.warning('StackdriverExporter set up failed. Using FileExporter.')
+        exporter = file_exporter.FileExporter(
+            transport=background_thread.BackgroundThreadTransport)
+        LOGGER.info(
```
Right.

blueandgold commented on google/cloud/forseti/services/client.py:
```diff
@@ -718,7 +718,9 @@ def __init__(self, endpoint=DEFAULT_ENDPOINT, ping=False):
         """
-        self.channel = grpc.insecure_channel(endpoint)
+        self.endpoint = endpoint
+        self.channel = grpc.insecure_channel(self.endpoint)
```
Yes, your proposal is much more clear for me:
```
self.channel = grpc.intercept_channel(
    grpc.insecure_channel(endpoint),
    *self.interceptors)
```

blueandgold commented on google/cloud/forseti/services/tracing.py:
```diff
@@ -0,0 +1,84 @@
+        LOGGER.info(
+            'StackdriverExporter set up successfully for project %s.',
+            exporter.project_id)
+    except Exception as e:
```
We actually don't mind broad exceptions as a mean to keep the server from fatal crashes, and the exception event is handled appropriately and logged with stacktraces.

ocervell commented:
Yes, we need tests at least for the functions included in `tracing.py`.
Changed title to WIP since there is [a bug](https://github.com/census-instrumentation/opencensus-python/issues/270) blocking this to be merged.

- [ ] Add unit tests
- [ ] Include integration libraries tracing (requests, httplib, sqlalchemy)
- [ ] [Bugfix] Stackdriver exporter crash handler for OpenCensus

ocervell commented on google/cloud/forseti/common/opencensus/tracing.py:
```diff
@@ -0,0 +1,108 @@
+        exporter)
+
+
+def trace_integrations(integrations=DEFAULT_INTEGRATIONS):
```
isn't 'trace' a verb ? Like 'trace those integrations' ?

ocervell replied:
Would it be better called `trace_integration_libraries` ?

ocervell commented on google/cloud/forseti/services/client.py:
```diff
@@ -69,6 +70,24 @@ def wrapper(*args, **kwargs):
+    if is_opencensus_enabled():
+        # It's okay for this to be enabled on the client, even if the tracing
+        # flag is disabled on the server.
+        from google.cloud.forseti.common.opencensus import tracing
```
We have to make an exception here: if we put the import at the top of the file, the import would fail if tracing is disabled because `opencensus` is not even installed. 

That's why the code checks if `is_opencensus_enabled()` before running this import, otherwise we would fail with an `ImportError`. 

I find this more elegant than `try`/`except` blocks for imports in `server.py` and `client.py`, what do you think ?

ocervell replied:
Sure, but keep in mind that this means wherever we'll need to add spans to traces (in each different service, such as inventory), we'll have to put the same try / except at the top of each file that we add tracing too.
Is this something acceptable ?

ocervell commented on tests/common/opencensus/tracing_test.py:
```diff
@@ -0,0 +1,89 @@
+    @mock.patch('opencensus.trace.ext.grpc.client_interceptor.OpenCensusClientInterceptor')
+    def test_create_client_interceptor(self, mock):
+        create_client_interceptor('localhost')
+        self.assertTrue(mock.called)
```
I need to check if the OpenCensusClientInterceptor can be instantiated without a gRPC server running.

blueandgold commented on google/cloud/forseti/common/opencensus/tracing.py:
```diff
@@ -0,0 +1,108 @@
+    exporter = create_exporter()
+    sampler = always_on.AlwaysOnSampler()
+    if extras:
+        trace_integrations()
```
Would it be very hard to assert on the state of `config_integration` after `config_integration.trace_integrations()` is called?  I would think that it should be possible.

If that is too hard, can you please put a comment on the return statement in that the returned values are only used for testing?  So that in the future we can remember the return is there for testing purposes.

ocervell commented:
Thanks for your in-depth review. I think I've addressed the above comments with my last commits.
